File: Assets/Bundles/Shared/Resources/TS/VoxelWorld/VoxelData/VoxelDataAPI.lua
-- Compiled with unity-ts v2.1.0-75
local Network = require("Shared/TS/Network").Network
local RunUtil = require("Shared/TS/Util/RunUtil").RunUtil
local VoxelDataAPI
do
	VoxelDataAPI = setmetatable({}, {
		__tostring = function()
			return "VoxelDataAPI"
		end,
	})
	VoxelDataAPI.__index = VoxelDataAPI
	function VoxelDataAPI.new(...)
		local self = setmetatable({}, VoxelDataAPI)
		return self:constructor(...) or self
	end
	function VoxelDataAPI:constructor()
	end
	function VoxelDataAPI:Init()
		if RunCore:IsClient() then
			Network.ServerToClient.SetVoxelData.Client:OnServerEvent(function(voxelPos, key, data)
				self:SetVoxelData(voxelPos, key, data)
			end)
		end
	end
	function VoxelDataAPI:SetVoxelData(voxelPos, key, data)
		local map
		local _voxelDataMap = self.voxelDataMap
		local _voxelPos = voxelPos
		if _voxelDataMap[_voxelPos] ~= nil then
			local _voxelDataMap_1 = self.voxelDataMap
			local _voxelPos_1 = voxelPos
			map = _voxelDataMap_1[_voxelPos_1]
		else
			map = {}
			local _voxelDataMap_1 = self.voxelDataMap
			local _voxelPos_1 = voxelPos
			local _map = map
			_voxelDataMap_1[_voxelPos_1] = _map
		end
		local _map = map
		local _key = key
		local _data = data
		_map[_key] = _data
		if RunUtil:IsServer() then
			Network.ServerToClient.SetVoxelData.Server:FireAllClients(voxelPos, key, data)
		end
	end
	function VoxelDataAPI:ClearVoxelData(voxelPos)
		local _voxelDataMap = self.voxelDataMap
		local _voxelPos = voxelPos
		local _result = _voxelDataMap[_voxelPos]
		if _result ~= nil then
			table.clear(_result)
		end
	end
	function VoxelDataAPI:GetVoxelData(voxelPos, key)
		local _voxelDataMap = self.voxelDataMap
		local _voxelPos = voxelPos
		local _result = _voxelDataMap[_voxelPos]
		if _result ~= nil then
			local _key = key
			_result = _result[_key]
		end
		return _result
	end
	function VoxelDataAPI:GetParentVoxelPos(childPos)
		local _childVoxelRedirectMap = self.childVoxelRedirectMap
		local _childPos = childPos
		return _childVoxelRedirectMap[_childPos]
	end
	function VoxelDataAPI:GetChildrenVoxelPos(parentPos)
		local _parentToChildrenMap = self.parentToChildrenMap
		local _parentPos = parentPos
		local _condition = _parentToChildrenMap[_parentPos]
		if _condition == nil then
			_condition = {}
		end
		return _condition
	end
	function VoxelDataAPI:SetChildOfParent(childPos, parentPos)
		local _childVoxelRedirectMap = self.childVoxelRedirectMap
		local _childPos = childPos
		local _parentPos = parentPos
		_childVoxelRedirectMap[_childPos] = _parentPos
		local _parentToChildrenMap = self.parentToChildrenMap
		local _parentPos_1 = parentPos
		local set = _parentToChildrenMap[_parentPos_1]
		if not set then
			set = {}
			local _parentToChildrenMap_1 = self.parentToChildrenMap
			local _parentPos_2 = parentPos
			local _set = set
			_parentToChildrenMap_1[_parentPos_2] = _set
		end
		local _set = set
		local _childPos_1 = childPos
		_set[_childPos_1] = true
	end
	VoxelDataAPI.voxelDataMap = {}
	VoxelDataAPI.childVoxelRedirectMap = {}
	VoxelDataAPI.parentToChildrenMap = {}
end
return {
	VoxelDataAPI = VoxelDataAPI,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Typescript~/src/Shared/VoxelWorld/VoxelData/VoxelDataAPI.ts
import { Network } from "Shared/Network";
import { RunUtil } from "Shared/Util/RunUtil";

export class VoxelDataAPI {
	private static voxelDataMap = new Map<Vector3, Map<string, unknown>>();

	/**
	 * Some prefab blocks take up more than 1x1x1 space (for example: a bed.)
	 * This map is a child fake block that points to the source block position.
	 */
	private static childVoxelRedirectMap = new Map<Vector3, Vector3>();
	private static parentToChildrenMap = new Map<Vector3, Set<Vector3>>();

	public static Init(): void {
		if (RunCore.IsClient()) {
			Network.ServerToClient.SetVoxelData.Client.OnServerEvent((voxelPos, key, data) => {
				this.SetVoxelData(voxelPos, key, data);
			});
		}
	}

	public static SetVoxelData(voxelPos: Vector3, key: string, data: unknown): void {
		let map: Map<string, unknown>;
		if (this.voxelDataMap.has(voxelPos)) {
			map = this.voxelDataMap.get(voxelPos)!;
		} else {
			map = new Map<string, unknown>();
			this.voxelDataMap.set(voxelPos, map);
		}
		map.set(key, data);
		if (RunUtil.IsServer()) {
			Network.ServerToClient.SetVoxelData.Server.FireAllClients(voxelPos, key, data);
		}
	}

	public static ClearVoxelData(voxelPos: Vector3): void {
		this.voxelDataMap.get(voxelPos)?.clear();
	}

	public static GetVoxelData<T>(voxelPos: Vector3, key: string): T | undefined {
		return this.voxelDataMap.get(voxelPos)?.get(key) as T | undefined;
	}

	public static GetParentVoxelPos(childPos: Vector3): Vector3 | undefined {
		return this.childVoxelRedirectMap.get(childPos);
	}

	public static GetChildrenVoxelPos(parentPos: Vector3): Set<Vector3> {
		return this.parentToChildrenMap.get(parentPos) ?? new Set<Vector3>();
	}

	public static SetChildOfParent(childPos: Vector3, parentPos: Vector3): void {
		this.childVoxelRedirectMap.set(childPos, parentPos);
		let set = this.parentToChildrenMap.get(parentPos);
		if (!set) {
			set = new Set<Vector3>();
			this.parentToChildrenMap.set(parentPos, set);
		}
		set.add(childPos);
	}
}
